### PR TITLE
core: recover the state in SetChainHead if the head state is missing

### PR DESCRIPTION
**Pull request by rjl493456442**
This PR improves the `SetChainHead`, with the additional guarantee that after head reset the head state is always available.

### CODE REVIEW

File: core/blockchain.go [hunk 1]
```diff
@@ -2106,6 +2106,14 @@ func (bc *BlockChain) SetChainHead(head *types.Block) error {
 	}
 	bc.writeHeadBlock(head)
 
+	// Re-execute the reorged chain in case the head state is missing.
+	// It should be done before firing events.
+	if !bc.HasState(head.Root()) {
+		if err := bc.recoverAncestors(head); err != nil {
+			return err
```
Comment (holiman) [inline core/blockchain.go:2113]: If we wind up here, we have already done `bc.writeHeadBlock(head)`. Isn't that a problem? And wouldn't it be better if we do this check/recovery a bit earlier in this method?

Comment (rjl493456442) [reply core/blockchain.go]: You are right, otherwise we might mess up the RPC, fixed it.

**Comment by fjl** [comment]
@MariusVanDerWijden Please confirm that this change matches your intuition about how the ForkchoiceUpdated API should work.

**Comment by MariusVanDerWijden** [comment]
One question I have is that since it might take a long time to recover the state if an attacker could somehow trigger all nodes on the network to recover an old state and thus effectively force all nodes to stop block processing

**Comment by rjl493456442** [comment]
@MariusVanDerWijden If it happen(deep reorg), then we need to recover the state anyway, no matter in this call `SetHead` or when insert the next block.

It depends on how many blocks we need to reorg, but I guess, it can be limited by the PoS itself(finalization, whatever else), we can't help too much in this case.

**Review by MariusVanDerWijden — APPROVED** [review]
LGTM

**Review by karalabe — APPROVED** [review]
SGTM

**Pull request merged**